### ISSUE TITLE
register_hook-modified gradient cannot be applied to optimizer

### ISSUE THREAD

**Issue by tailintalent** [pytorch/pytorch#2843]
I want to add some Gaussian noise to the gradient before passing the gradient to the optimizer. I tried using register_hooks, as the test code below:

```
X = Variable(torch.FloatTensor([4]))
a = Variable(torch.FloatTensor([2]), requires_grad = True)
y = a * X
criterion = nn.MSELoss(size_average = True)
optimizer = torch.optim.Adam([a], lr = 1e-2)

h = a.register_hook(lambda grad: grad * 10)
optimizer.zero_grad()
loss = criterion(y, Variable(torch.FloatTensor([7])))
loss.backward()
optimizer.step()
print(a)
print(a.grad)
```
But I find that although I can modified the gradient at will, the optimizer always updates the gradient using the original gradient, not the one after register_hook(). Is this behavior what is supposed to be? Or how can I applied the modified gradient to the optimizer?

Thanks!

**Comment by chenzhekl** [comment]
Your script outputs `80` for `a.grad` which is correct. What is your issue?

**Comment by tailintalent** [comment]
The a.grad is correct. My issue is, I expect that if I change the gradient, the variable "a" after the optimizer.step() should also change. However, using the above code, after one backward step, a decreases from 2 to 1.9900. If I change this line:
`h = a.register_hook(lambda grad: grad * 10)`
to
`h = a.register_hook(lambda grad: grad * 10000)`

After backward step the variable "a" still decrease from 2 to 1.9900. Therefore, it shows that my manipulation of the gradient has no effect on the optimizer. Is that the supposed behavior? On the other hand, I want the optimizer to take in my manipulated gradient, not the original gradient. How should I do that?

**Comment by sniklaus** [comment]
Try using SGD and you will see the behavior that you are expecting.

**Comment by tailintalent** [comment]
Is there a way that Adam and other non-SGD optimizers can do that? Or should I write my own optimizer to achieve that?

**Comment by chenzhekl** [comment]
Yes, it is expected. If your look at [the formula of Adam](https://arxiv.org/pdf/1412.6980.pdf), you'll find that Adam is scaling invariant.  It means that if we have some objective function f(x) and we change it to k*f(x) (where k is some constant), there will be no effect on performance. What you did is the same as scaling the objective function (since you are requiring and modifying just `a`).

**Comment by tailintalent** [comment]
Thanks!

**Comment by soumith** [comment]
that's quite a good analysis and explanation @chenzhekl :)
